Add restrospetive KEP for ProcMount and target at Beta (#4266)
kubernetes/enhancements: keps/sig-node/4265-proc-mount/README.md

### Non-Goals

## Proposal

Add a new `string` type field named `ProcMountType` will hold the viable
options for `procMount` to the `SecurityContext`
definition.

By default,`procMount` is `default`, aka the same behavior as today and the
paths are masked.

This will look like the following in the spec:

```go
type ProcMountType string

const (
    // DefaultProcMount uses the container runtime default ProcType.  Most 
    // container runtimes mask certain paths in /proc to avoid accidental security
    // exposure of special devices or information.
    DefaultProcMount ProcMountType = "Default"

    // UnmaskedProcMount bypasses the default masking behavior of the container
    // runtime and ensures the newly created /proc the container stays in tact with
    // no modifications.  
    UnmaskedProcMount ProcMountType = "Unmasked"
)

procMount *ProcMountType
```

This requires changes to the CRI runtime integrations so that kubelet will add the specific `unmasked` option.

The main use case for unmasking paths in `/proc` are for a user nesting unprivileged containers within a container. However, having an Unmasked ProcMountType

> is a privileged operation, and thus is part of the [privileged PSS policy](https://github.com/kubernetes/kubernetes/pull/103340). Since a user must have be in

Reviewer: this PR says baseline

Author: yeah  I was a bit confused at first too, but https://github.com/kubernetes/kubernetes/pull/103340/files#diff-88fde5f23fad7b65d37f3783e7980b18e369b6e03f22c0575611628eec7f476dR63 shows it will fail admission if the value is anything other than default or empty, which explicitly disallows Unmasked for baseline and under

  - Components depending on the feature gate: kubelet, kube-apiserver

###### Does enabling the feature change any default behavior?

No, only allows a user to access the Unmasked ProcMountType

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes

###### What happens if we reenable the feature if it was previously rolled back?

Nothing unexpected.

###### Are there any tests for feature enablement/disablement?

No

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

Inconsistent feature gate enablement is the only way. All supported versions of CRIs (CRI-O and containerd) have support for this feature.

###### What specific metrics should inform a rollback?

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

They will be, specifically cases where the Kubelet is restarted.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

> Can check feature gates enabled on kubelet and kube-apiserver.

Reviewer: This means its enabled, doesn't mean its in use.  Is it a query of some kind of pod?  Does the kubelet have a metric?

Author: updated with a kubectl query!

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [x] Other (treat as last resort)

>   - Details: Container created with the Unmasked ProcMountType have paths [here](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L193) as writable, not read only.

Reviewer: nit: explode list here please.

PSS has already been configured to mitigate this by required a user be in a privileged namespace to get access to the field.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

No effect

###### What are other known failure modes?

Malicious user, protected by PSS.

###### What steps should be taken if SLOs are not being met to determine the problem?

The field can be unset in a pod spec to see if SLOs met after the feature is disabled.

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

2018-05-07: k/community update opened
2018-05-27: k/kubernetes PR merged with support.
2023-10-02: KEP opened and targeted at Beta

## Drawbacks

> ## Alternatives

Reviewer 2: Has anyone tried to submit a patch to the kernel to allow mounting a new procfs with masks?

Author: that's a cool idea! I have no idea if it's been tried or how difficult it would be

Reviewer 3: cc: @giuseppe

Author: added as an alternative

Reviewer 4: I'm not aware of any effort to allow it.  AFAIK the risk is that allowing a full proc mount will enable a user to access the previously masked files.

This should be safer together with user namespaces, at least it will be possible to allow nested containers in the userns limited environment.

Tagging @brauner in case he has better news/ideas from the kernel side :-)

Reviewer 5: The response was no in 2018: https://patchwork.kernel.org/project/linux-fsdevel/patch/20180404115311.725-1-alban@kinvolk.io/ -- I don't think much has changed really.

I do believe something similar to the LXD approach (I mention below) could be an idea with the advantage it can apply to /sys too.

- [x] (R) Production readiness review completed
- [x] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Currently the most container runtimes work is by masking and setting as read-only certain paths in `/proc`.
This is to prevent data from being exposed into a container that should not be.
However, there are certain use-cases where it is necessary to turn this off.

> In 1.12, this was introduced as the ProcMountType feature gate, and has languished since. This KEP is

Reviewer 6: ```suggestion
In 1.12, this was introduced as the ProcMountType feature gate, and it has languished in alpha ever since. This KEP is
```

> a successor to (and heavily based on) https://github.com/kubernetes/community/pull/1934/, updated for the modern era.

Author: This is meant to be more of a citation than a summary

> a successor to (and heavily based on) https://github.com/kubernetes/community/pull/1934/, updated for the modern era.

Reviewer 7: For a KEP, we typically put an actual summary here. That's what I'm recommending. You could keep the “In 1.12, this was introduced” short history if the text before it was a summary of the proposal.

(right now, the summary gives context and history but does not **propose** any change)

## Motivation

> For end-users who would like to run unprivileged containers using user namespaces _nested inside_ CRI containers, we need an option to have a `ProcMount`.

Reviewer 7: ```suggestion
Some end users would like to run unprivileged containers that are using user namespaces _nested inside_ containers that are defined in a Pod. The outer containers are visible to the kubelet and, through CRI,
to the container runtime. The kubelet is unaware of the nested inner containers.
Kubernetes defaults to masking the proc mount and setting some paths readonly. When end users want
to run nested containers inside an unprivileged Pod, those users want a way to override that default
behavior.
```

> That is, we need an option to designate explicitly turn off masking and setting read-only of paths so that we can mount `/proc` in the nested container as an unprivileged user.

Reviewer 6: ```suggestion
That is, we need an option to explicitly turn off masking and setting read-only of paths in `/proc` to allow mounting `/proc` in the nested container as an unprivileged user.
```

Please see the following filed issues for more information:
- [opencontainers/runc#1658](https://github.com/opencontainers/runc/issues/1658#issuecomment-373122073)
- [moby/moby#36597](https://github.com/moby/moby/issues/36597)
- [moby/moby#36644](https://github.com/moby/moby/pull/36644)

### Goals

> - Add the ProcMountType option to the pod API, allowing users to override the proc mount behavior in a container.

Reviewer 7: ```suggestion
- Allow users to customize the proc mount behavior for a Linux container.
```

### Non-Goals

## Proposal

> Add a new `string` type field named `ProcMountType` will hold the viable

Reviewer 6: ```suggestion
Add a new `string` type named `ProcMountType` for choosing from a set of proc mount isolation mode options.  These
```

> options for `procMount` to the `SecurityContext`

Reviewer 6: ```suggestion
options for `procMount` will be introduced as a new field in the `SecurityContext` (LinuxContainerSecurityContext/WindowsContainerSecurityContext)
```

definition.

> By default,`procMount` is `default`, aka the same behavior as today and the

Reviewer 6: ```suggestion
The default for `procMount` (`default`) allows the container runtime to configure it's own defaults, for example as is done today with the result being the 
```

Author: the paths are actually passed from the kubelet today. I have adopted some of your wording but kept that impl detail

paths are masked.

This will look like the following in the spec:

```go
type ProcMountType string

const (
    // DefaultProcMount uses the container runtime default ProcType.  Most 
    // container runtimes mask certain paths in /proc to avoid accidental security
    // exposure of special devices or information.

>     DefaultProcMount ProcMountType = "Default"

Reviewer 6: case?

Author: this is what is there today

    // UnmaskedProcMount bypasses the default masking behavior of the container

>     // runtime and ensures the newly created /proc the container stays in tact with

Reviewer 6: ```suggestion
    // runtime and ensures the newly created /proc the container stays intact with
```

    // no modifications.  
    UnmaskedProcMount ProcMountType = "Unmasked"
)

procMount *ProcMountType
```

where nil is default, and is interpreted as "Default" ProcMountType.

When the kubelet is presented with a pod that has a ProcMountType as Unmasked, it will edit the default list of

> masked paths it passes down to the [CRI to be empty](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L216) which it does

Reviewer 7: (nit)
```suggestion
masked paths it passes down to the CRI to be [empty](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L216) which it does
```
“CRI to be empty” is an awkward phrase to use as a hyperlink

with the [CRI request](https://github.com/kubernetes/kubernetes/blob/964529b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L889-L891).

This requires changes to the CRI runtime integrations so that kubelet will add the specific `unmasked` option.
This was done after alpha:

> CRI-O has support in v1.25.0 after https://github.com/cri-o/cri-o/pull/6025/commits/4102586132214263c5d0ae93ec257432653ab82b
> containerd has support in 1.6. See https://github.com/containerd/containerd/pull/5070/commits/07f1df4541d6a81c205d194f4f6ea3a6a95c3e29

Reviewer 7: ```suggestion
- CRI-O has support in v1.25.0 after https://github.com/cri-o/cri-o/pull/6025/commits/4102586132214263c5d0ae93ec257432653ab82b
- containerd has support in 1.6. See https://github.com/containerd/containerd/pull/5070/commits/07f1df4541d6a81c205d194f4f6ea3a6a95c3e29
```

The main use case for unmasking paths in `/proc` are for a user nesting unprivileged containers within a container. However, having an Unmasked ProcMountType

> is a privileged operation, and thus is part of the [privileged PSS policy](https://github.com/kubernetes/kubernetes/pull/103340). Since a user must have be in

Reviewer 7: ```suggestion
is a privileged operation, and thus is part of the [privileged](https://k8s.io/docs/concepts/security/pod-security-standards/#privileged) Pod Security Standard. Since a user must have be in
```
PodSecurity<i>Policies</i> are no longer part of Kubernetes

Reviewer 7: (addressed)

> the privileged policy, they are also trusted to choose the correct user ID and run a workload that won't interfere with the host.

Reviewer: This statement appears to be incorrect.  It looks like procMount can be set in baseline: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/pod-security-admission/policy/check_procMount.go#L47-L58

Is it incorrectly categorized?

> A container running as root user on the host and an unmasked `/proc` could be able to write to the host `/proc`, and thus this privileged designation is appropriate.

Reviewer: Pretty sure we allow baseline to do this today.

Author: Though baseline doesn't get access to unmasked proc mount, so the only way a user can get access to both is privileged

> ### User Stories (Optional)

Reviewer 5: There are potentially some other (very) niche user stories, e.g. key rings are namespaced correctly in user namespaces, so it doesn't make sense to limit access to `/proc/keys` (the user story here is running some credential storage tools that use `keyrings(7)` and wanting to debug them).

#### Story 1

As a cluster admin, I would like a way to nest containers within containers. To do so, kernel the top level containers need an unmasked /proc.

#### Story 2

As a kubernetes user, I may want to build containers from within a kubernetes container. 
See [this article for more information](https://github.com/jessfraz/blog/blob/master/content/post/building-container-images-securely-on-kubernetes.md).

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

> How will security be reviewed, and by whom?

Reviewer 5: A user turning this on without user namespaces enforced means root within a container can trivially escape. Pod Security Standards obviously enforces, but worth noting.

Reviewer 7: I guess you mean that enforcing the _restricted_ Pod security standard prevents that escape?

> How will security be reviewed, and by whom?

Reviewer 5: Risk: Placing more trust in the user namespacing mechanism of the kernel. (Arguably the kernel being responsible for making /proc and /sys secure is better anyway, just worth explicitly drawing attention to.)

-->

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

##### Unit tests

- `pkg/securitycontext`: `10-05-2023` - `70.04`

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

> - N/A

Reviewer 7: Why are these not applicable?

##### e2e tests

- test/e2e_node

- additional tests should be added to e2e_node suite to test the adherence of the ProcMount field
  - Test default behavior actually masks /proc paths.
  - Test Unmasked behavior is not masking /proc paths.
  - Test PSS policies (if possible to test in e2e)

>   - Test that Windows pod cannot be scehduled with the value of ProcMount specifies

Author: yeah

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->
Turn off the feature gate to turn off the feature.

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and nodes?
- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

> It has been in support by kube-apiserver and kubelet since 1.12, so version skew is no longer a worry.

Reviewer: is the kubelet functionality featuregated?  if so, are there safety concerns when used outside of combination with user namespaces that require coordinated rollout?

Author: actually it turns out no. just KAS

> It has been in support by kube-apiserver and kubelet since 1.12, so version skew is no longer a worry.

Reviewer: If the kube-apiserver allows a field to be set and the kubelet has the gate unset, what happens?

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ProcMountType
  - Components depending on the feature gate: kubelet, kube-apiserver

###### Does enabling the feature change any default behavior?

No, only allows a user to access the Unmasked ProcMountType

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes

Reviewer: What is is disabled, how do I actually stop pods with this access?  I'm better we have to drain nodes and that needs to be described here.

###### What happens if we reenable the feature if it was previously rolled back?

> Nothing unexpected.

Reviewer: when re-enabled, you have to evict pods again because the existing pods won't actually have their intent fulfilled, right?

No

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

Inconsistent feature gate enablement is the only way. All supported versions of CRIs (CRI-O and containerd) have support for this feature.

###### What specific metrics should inform a rollback?

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

They will be, specifically cases where the Kubelet is restarted.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

`kubectl get pods --all-namespaces -o jsonpath="{range .items[*]}{.metadata.name}{' '}{.spec.containers[*].securityContext.procMount}{'\n'}{end}"  | grep -i unmasked`
Will print all pods that has an Unmasked ProcMountType, along with the pod name.

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [x] Other (treat as last resort)

>   - Details: Container created with the Unmasked ProcMountType have paths [here](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L193) as writable, not read only.

Reviewer: if the umasked proc mount fails, will the pod fail to start and produce a good status message?

Author: This kind of error condition would be bubbled up from the runtime presumably, so yes

Author: hopefully it's good...

Reviewer: > This kind of error condition would be bubbled up from the runtime presumably, so yes

How about a CRIO unit test to check?

No

###### Will enabling / using this feature result in introducing new API types?

ProcMountType in the pod spec

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

Negligible

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

Potentially a malicious user given access and running with a root container in the host context could mess with the host processes.
PSS has already been configured to mitigate this by required a user be in a privileged namespace to get access to the field.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

No effect

###### What are other known failure modes?

> Malicious user, protected by PSS.

Reviewer 7: (addressed)

###### What steps should be taken if SLOs are not being met to determine the problem?

The field can be unset in a pod spec (or feature gate turned off) to see if SLOs met after the feature is disabled for pods.

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

2018-05-07: k/community update opened
2018-05-27: k/kubernetes PR merged with support.
2023-10-02: KEP opened and targeted at Beta

## Drawbacks

> ## Alternatives

Reviewer 5: LXD uses an approach where it mounts an additional `/dev/.lxc/proc` and `/dev/.lxc/sys`, the advantage being the kernel sees an unmasked proc filesystem (it's fine it's mounted anywhere, not just the root). It then uses AppArmor to deny user access to the paths -- meaning it can offer this feature (`security.nesting=true` in LXD terms) regardless of what paths are masked.

I believe an approach like this has benefits -- it means that /sys can also be mounted in unprivileged containers (because usually `/sys` is masked by `/sys/fs/cgroup` anyway). I begun writing a KEP targeting 1.30 (I haven't published it yet, was waiting for 1.29 KEP to be over, as I wasn't expecting this to get rushed at the last minute).

Author: > (it's fine it's mounted anywhere, not just the root).

this is a weird limitation... thanks for the context, I'll add the alternative, and we can chat about alternatives. it's possible the runtimes could do something fancy like that instead of masking...

    // container runtimes mask certain paths in /proc to avoid accidental security
    // exposure of special devices or information.
    DefaultProcMount ProcMountType = "Default"

    // UnmaskedProcMount bypasses the default masking behavior of the container
    // runtime and ensures the newly created /proc the container stays in tact with
    // no modifications.  
    UnmaskedProcMount ProcMountType = "Unmasked"
)

procMount *ProcMountType
```

This requires changes to the CRI runtime integrations so that kubelet will add the specific `unmasked` option.

Note: Pod Security Admission will also need to be taught how to handle this option.
[Luckily, this has been done after the feature was introduced.](https://github.com/kubernetes/kubernetes/pull/103340)

Unmasking the paths in `/proc` option really only makes sense for when a user is nesting unprivileged containers with user namespaces as it will allow more information
than is necessary to the program running in the container spawned by kubernetes.

It should be noted that this is different that the host /proc. It is still a newly mounted /proc just the container runtimes will not mask the paths.

> Since the only use case for this option is to run unprivileged nested containers, this option should only be allowed or used if the user in the container is not `root`.

Reviewer: should or must? If must, it will be enforced at API validation?

Author: actually I've just checked--only privileged users can have access to this feature. I have left a note on the dangers of allowing users to access this field. I think a future enhancement could be to allow non root users access to an unmasked proc in the baseline policy (especially if they're confined by a user namespace), but more research has to happen for that.

Reviewer: so it's actually the exact opposite of what this sentence said this morning?

Author: yeah I misread the code

This can be easily enforced with `MustRunAs`.

> Since the user inside is still unprivileged, doing things to `/proc` would be off limits regardless, since linux user support already prevents this.

Reviewer: is "doing things" reads or just writes?

Author: just writes

### User Stories (Optional)

#### Story 1

> As a cluster admin, I would like a way to nest containers within containers and get access to proc entries as if the top level container was running

Reviewer 2: AFAICS you are not interested in getting access to those masked procfs entries.
The actual reason you want to unmask them is because the kernel doesn't allow mounting a new procfs without unmasking them, no?

in the host context.

#### Story 2

> As a cluster admin, I may want to [genuinetools/img](https://github.com/genuinetools/img) inside a kubernetes container.

Reviewer 2: genuinetools/img is no longer maintained.
https://github.com/genuinetools/img/issues/348#issuecomment-982473254

Author: dropped!

That program then launches sub-containers that take advantage of user namespaces and re-mask /proc and set /proc as read-only.

#### Story 3

As a kubernetes user, I may want to build containers from within a kubernetes container. 

> See [this article for more information](https://github.com/jessfraz/blog/blob/master/content/post/building-container-images-securely-on-kubernetes.md).

Reviewer 2: Story 3 seems to overlap with Story 2

Author: yeah I dropped 2

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ProcMountType
  - Components depending on the feature gate: kubelet, kube-apiserver

###### Does enabling the feature change any default behavior?

No, only allows a user to access the Unmasked ProcMountType

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes

###### What happens if we reenable the feature if it was previously rolled back?

Nothing unexpected.

###### Are there any tests for feature enablement/disablement?

No

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

Inconsistent feature gate enablement is the only way. All supported versions of CRIs (CRI-O and containerd) have support for this feature.

###### What specific metrics should inform a rollback?

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

> No, though since its a fairly contained feature I would be surprised if there are issues.

Reviewer: Please commit to a manual check anyway.  Particularly cases where kubelets are restarted without interrupting existing containers.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

Can check feature gates enabled on kubelet and kube-apiserver.

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [x] Other (treat as last resort)

>   - Details: successfully create a pod with a container that uses ProcMountType

Reviewer: Behavior silently drops fields, so it won't actually fail.  What's actually needed?  Create of procmounttype and then an initContainer that actually checks the filesystem?

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

N/A

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [ ] Metrics
  - Metric name:
  - [Optional] Aggregation method:
  - Components exposing the metric:
- [x] Other (treat as last resort)

>   - Details: general health of pods that use this feature.

Reviewer: name a specific metric to watch for a change before/after.

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

- A CRI implementation that supports this feature
    - All supported versions currently do.

### Scalability

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

ProcMountType in the pod spec

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

Negligible

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

> Potentially a malicious user given access and running with a root container in the host context could mess with the host processes, but PSA has already

Reviewer: Fix how?  There's a "should" above that isn't clear about should or must. And the consequences here are not obvious to me.  And if should isn't enforced today in beta and really ought to be must, then additional consideration/work is required.  We could discuss breaking options since its only alpha today.

Author: is the feature being restricted to users in a privileged PSA sufficient to mitigate? I feel like it is

been configured to fix this.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

No effect

###### What are other known failure modes?

Malicious user, protected by PSA.

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

2018-05-07: k/community update opened
2018-05-27: k/kubernetes PR merged with support.

> 2023-10-02: KEP opened and targeted at Beta

Reviewer 2: What was the story in 2018-2023?

## Drawbacks

> ## Alternatives

Reviewer 2: What about `--oci-worker-no-process-sandbox` ?
https://github.com/moby/buildkit/blob/v0.12.2/examples/kubernetes/job.rootless.yaml#L31

a successor to (and heavily based on) https://github.com/kubernetes/community/pull/1934/, updated for the modern era.

## Motivation

For end-users who would like to run unprivileged containers using user namespaces _nested inside_ CRI containers, we need an option to have a `ProcMount`.
That is, we need an option to designate explicitly turn off masking and setting read-only of paths so that we can mount `/proc` in the nested container as an unprivileged user.

Please see the following filed issues for more information:
- [opencontainers/runc#1658](https://github.com/opencontainers/runc/issues/1658#issuecomment-373122073)
- [moby/moby#36597](https://github.com/moby/moby/issues/36597)
- [moby/moby#36644](https://github.com/moby/moby/pull/36644)

### Goals

- Add the ProcMountType option to the pod API, allowing users to override the proc mount behavior in a container.

### Non-Goals

## Proposal

Add a new `string` type field named `ProcMountType` will hold the viable
options for `procMount` to the `SecurityContext`
definition.

By default,`procMount` is `default`, aka the same behavior as today and the

> paths are masked.

Reviewer 8: should we have Masked as an explicit option? If this wasn't a part of an API before, making an explicit Masked option would be a better design

Reviewer 8: it will ensure that the user may force masking as a security feature

Author: I am less inclined personally--feels like syntactic sugar. Functionally, default is masked, and a use could just as easily enforce default (see PSS today) as masked

Reviewer 8: discussed with @mrunalp - it may be a good idea to change the descriprion of "default" to say it means that certain subpaths are masked. Link to or oci list https://github.com/kubernetes/kubernetes/blob/598076a71caeaf7720b758eedd5cca5a6364ef62/pkg/securitycontext/util.go#L193-L211

If default meaning is "whatever", I wonder if we can even write an e2e test validating any behavior for the "default"

Author: +1 to clarifying the definition: would you like me to capture that work here, and how? 

We can test what default means because it's interpreted by the kubelet, so it's all here. A better situation than if we passed the proc mount type to CRI and then it decided what to do with the field. We can test for the masked value of the files in the container's rootfs.

Reviewer 8: yes, let's mention that we will clarify the semantic of this field

Reviewer 6: see comment ^ IMO default allows the container runtime to configure .. much as we handle other default isolation tooling options..

This will look like the following in the spec:

```go
type ProcMountType string

const (
    // DefaultProcMount uses the container runtime default ProcType.  Most 
    // container runtimes mask certain paths in /proc to avoid accidental security
    // exposure of special devices or information.
    DefaultProcMount ProcMountType = "Default"

    // UnmaskedProcMount bypasses the default masking behavior of the container
    // runtime and ensures the newly created /proc the container stays in tact with
    // no modifications.  
    UnmaskedProcMount ProcMountType = "Unmasked"
)

> procMount *ProcMountType

Reviewer 8: `nil` is default?

Also current documentaiton is saying: "// Note that this field cannot be set when spec.os.name is windows."

Can we update this by specifying what to expect when it is Windows. Which component will blow out and what is expected Runtime reaction on this

Reviewer 8: this is nit for the implementation. Also would be good to add a test for Windows setting this value

Author: agreed

```

> This requires changes to the CRI runtime integrations so that kubelet will add the specific `unmasked` option.

Reviewer 8: are we ready to bump to beta if more changed on runtime is needed?

Author: I wrote this KEP as if it was written originally in alpha--this has been done with all supported versions of containerd and cri-o

Reviewer 8: if not needed - can you please list the supported versions.

Reviewer 6: for `unmasked` does kubelet need a way to (outside mounting..) to figure out this option has been implemented for the container?  Today we only return container id in response to create.. As an alternative we could add a response indicating unmasked is supported, or add a runtime feature.. or put the result / impl detail in info for container status...

[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Feature implemented behind a feature flag

#### Beta

- Add e2e tests for the feature

> - Investigate updates to PSS policies in conjunction with user namespaces

Reviewer 8: both criteria feels to be an alpha scope.

Author: PSS policies are already in place--just not with user namespaces. Agreed on e2e test, though I see no reason to block beta because they were not covered before

Reviewer 5: +1, I feel pushing this to beta is a bit early when the only safe way to use it is with user namespaces, which are still alpha (it is possible to use it with CRI-O annotations, which we have been doing, but also using that safely has some security footguns).

(cc @rata and @saschagrunert; I believe this could fit the PSS work in https://github.com/kubernetes/kubernetes/pull/118760)

Reviewer: Security posture and enforcement must be settled for alpha.

#### GA

- Allowing time for feedback

> - Investigate updates to PSS policies in conjunction with user namespaces

Reviewer 3: I would move this up to beta.

Author: good idea, updated

  - [x] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [x] (R) Graduation criteria is in place
  - [x] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [x] (R) Production readiness review completed
- [x] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> Currently most container runtimes mask and set as read-only certain paths in `/proc`.

Reviewer 7: ```suggestion
For Linux containers, common container runtimes mask and set as read-only certain paths in `/proc`.
```

This is to prevent data from being exposed into a container that should not be.
However, there are certain use-cases where it is necessary to turn this off.

> 

Reviewer 7: ```suggestion
This KEP proposes adding a field to the Pod security context to allow bypassing the usual restrictions.

```

In 1.12, this was introduced as the ProcMountType feature gate, and has it has languished in alpha ever since. This KEP is
a successor to (and heavily based on) https://github.com/kubernetes/community/pull/1934/, updated for the modern era.

## Motivation

Some end users would like to run unprivileged containers _nested inside_ a Kubernetes container using user namespaces. The outer container is started by the CRI implementation.
Kubernetes defaults to masking the `/proc` mount of a container, setting some paths as read only. To run a nested container within an unprivileged Pod, a user would need a way to
override that default masking behavior.

Please see the following filed issues for more information:
- [opencontainers/runc#1658](https://github.com/opencontainers/runc/issues/1658#issuecomment-373122073)
- [moby/moby#36597](https://github.com/moby/moby/issues/36597)
- [moby/moby#36644](https://github.com/moby/moby/pull/36644)

### Goals

- Allow users to opt out of the CRI masking `/proc` for Linux containers.

### Non-Goals

## Proposal

> Add a new `string` named `procMount` to the `securitContext` definition for choosing from a set of proc mount isolation mode options.

Reviewer 7: ```suggestion
Add a new `string` named `procMount` to the `securitContexty` definition for choosing from a set of proc mount isolation mode options.
```

> The default for`procMount` is `default`, which instructs the container runtime to mask the aforementioned paths.

Reviewer 7: ```suggestion
The default for`procMount` is `Default`, which instructs the container runtime to mask the aforementioned paths.
```

-->

##### Unit tests

- `pkg/securitycontext`: `10-05-2023` - `70.04`

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- N/A (Kubelet barely defines integration tests today, focusing on e2e_node tests instead)

##### e2e tests

- test/e2e_node

- additional tests should be added to e2e_node suite to test the adherence of the ProcMount field
  - Test default behavior actually masks /proc paths.
  - Test Unmasked behavior is not masking /proc paths.

>   - Test PSS integration (if possible to test in e2e)

Reviewer 3: Suggest a s/PSS/PSA/g in the doc.

[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Feature implemented behind a feature flag
- Add e2e tests for the feature (must be done before beta)

> - Investigate updates to PSA policies in conjunction with user namespaces

Reviewer 7: Do we need to add security warnings to the documentation for this feature? Maybe also the changelog for the first release after this KEP lands?

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ProcMountType
  - Components depending on the feature gate: kubelet, kube-apiserver (whole cluster must have `ProcMountType` feature gate enabled)

###### Does enabling the feature change any default behavior?

No, only allows a user to access the Unmasked ProcMountType

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes, though a reboot or drain will be necessary to limit any containers that originally got access to Unmasked ProcMountType.

###### What happens if we reenable the feature if it was previously rolled back?

The API server will need to recreate the pods that had the request scrubbed of the ProcMountType field (so the original intent can be fulfilled).

###### Are there any tests for feature enablement/disablement?

No

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

> Inconsistent feature gate enablement is the only way. All supported versions of CRIs (CRI-O and containerd) have support for this feature.

Reviewer 7: > All supported versions of CRIs (CRI-O and containerd) have support for this feature.

Does that include _all_ implementations of CRI, or only two widely used ones?

Reviewer 9: This is the type of problem I've mentioned above where you should prepare the feature to work when kube-apiserver or kubelet has it off.

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [x] Other (treat as last resort)
  - Details: Container created with the Unmasked ProcMountType have paths [here](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L193) as writable, not read only.
    - "/proc/asound",
	- "/proc/acpi",
	- "/proc/kcore",
	- "/proc/keys",
	- "/proc/latency_stats",
	- "/proc/timer_list",
	- "/proc/timer_stats",
	- "/proc/sched_debug",
	- "/proc/scsi",
	- "/sys/firmware",

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

N/A

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [x] Metrics
  - Metric name: `kubelet_pod_start_sli_duration_seconds`
  - [Optional] Aggregation method:
  - Components exposing the metric: kubelet
- [ ] Other (treat as last resort)
  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

- A CRI implementation that supports this feature

>     - All supported versions currently do.

Reviewer 7: Does that include all implementations of CRI?

Author: of containerd+ cri-o

Reviewer: update the KEP to be specific about which do.

ProcMountType in the pod spec

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

Negligible

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

Potentially a malicious user given access and running with a root container in the host context could mess with the host processes.
PSA has already been configured to mitigate this by required a user be in a privileged namespace to get access to the field.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

No effect

###### What are other known failure modes?

Malicious user gaining access to the host `/proc` with a rootful container

> - protected by Pod Security Admission, which requires the namespace is labeled in privileged PSA.

Reviewer 7: Not sure what this means. The default for Pod security admission is “no enforcement”.

###### What steps should be taken if SLOs are not being met to determine the problem?

The field can be unset in a pod spec (or feature gate turned off) to see if SLOs met after the feature is disabled for pods.

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

2018-05-07: k/community update opened
2018-05-27: k/kubernetes PR merged with support.

> 2023-10-02: KEP opened and targeted at Beta

Reviewer 3: Need to change this line.

Author: updated!

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

> [ ] I/we understand the owners of the involved components may require updates to

Reviewer 8: please check the box

existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

##### Unit tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

> - `<package>`: `<date>` - `<test coverage>`

Reviewer 8: any module you can think of posting here? The only think comes to mind is the file that setting values of security context. It actually has some decent coverage - I rememer more tests were added there recently.

Author: yeah good idea, added securitycontext

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

> - <test>: <link to test coverage>

Reviewer 8: I'd guess N/A

##### e2e tests

- test/e2e_node

> additional tests should be added to e2e_node suite to test the adherence of the ProcMount field field

Reviewer 8: ```suggestion
- additional tests should be added to e2e_node suite to test the adherence of the ProcMount field
  - Test default behavior actually masks /proc paths.
  - Test Unmasked behavior is not masking /proc paths.
  - Test PSS policies (can we even test it as e2e?)
  - Test that Windows pod cannot be scehduled with the value of ProcMount specifies
```

a successor to (and heavily based on) https://github.com/kubernetes/community/pull/1934/, updated for the modern era.

## Motivation

For end-users who would like to run unprivileged containers using user namespaces
_nested inside_ CRI containers, we need an option to have a `ProcMount`. That is,
we need an option to designate explicitly turn off masking and setting 
read-only of paths so that we can
mount `/proc` in the nested container as an unprivileged user.

Please see the following filed issues for more information:
- [opencontainers/runc#1658](https://github.com/opencontainers/runc/issues/1658#issuecomment-373122073)
- [moby/moby#36597](https://github.com/moby/moby/issues/36597)
- [moby/moby#36644](https://github.com/moby/moby/pull/36644)

Please also see the [use case for building images securely in kubernetes](https://github.com/jessfraz/blog/blob/master/content/post/building-container-images-securely-on-kubernetes.md).

Unmasking the paths in `/proc` option really only makes sense for when a user is nesting
unprivileged containers with user namespaces as it will allow more information
than is necessary to the program running in the container spawned by kubernetes.

> The main use case for this option is to run

Reviewer 3: May want to give other examples such as nested podman.

Author: I moved some stuff around to make the use cases clearer

- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

The feature gate is only processed by the API server--Kubelet has no awareness of it. API server will scrub the ProcMount field from the request
if it doesn't support the feature gate. Since all supported Kubelet versions support ProcMountType field, there's no version skew worry.
API server can have the feature gate toggled without worrying about doing the same for Kubelets.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ProcMountType
  - Components depending on the feature gate: kubelet, kube-apiserver (whole cluster must have `ProcMountType` feature gate enabled)

###### Does enabling the feature change any default behavior?

No, only allows a user to access the Unmasked ProcMountType

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes, though a reboot or drain will be necessary to limit any containers that originally got access to Unmasked ProcMountType.

Reviewer: how will a node reboot change the pod fields that are present?

Reviewer: Oh, i think I see.  Could you enumerate the steps more clearly

1. update featuregate on all kube-apiservers
2. restart all kube-apiservers
3. restart all nodes (to read the new pods, which will now have the fields stripped)

###### What happens if we reenable the feature if it was previously rolled back?

> The API server will need to recreate the pods that had the request scrubbed of the ProcMountType field (so the original intent can be fulfilled).

Reviewer: kube-apiserver doesn't create pods.  Which actor could/should do this?  Or perhaps nothing does?

Reviewer: Do workload resources also get their procMountType stripped when the featuregate is off?

Author: launching a cluster now to double check though

functionality is accessed.

[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Feature implemented behind a feature flag

> - Add e2e tests for the feature (must be done before beta)

Reviewer 9: If there exists tests, can you link them in the appropriate section above?

Author: none yet I will implement them before moving this to beta

  cluster required to make on upgrade, in order to make use of the enhancement?
-->
Turn off the feature gate to turn off the feature.

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and nodes?
- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

The feature gate is only processed by the API server--Kubelet has no awareness of it. API server will scrub the ProcMount field from the request
if it doesn't support the feature gate. Since all supported Kubelet versions support ProcMountType field, there's no version skew worry.

> API server can have the feature gate toggled without worrying about doing the same for Kubelets.

Reviewer 9: This is a bit confusing, the feature gate applies to both kube-apiserver and kubelet, in which case you should answer the question what happens when you turn the feature on one, and the other has it disabled?

Author: the feature gate only applies to KAS, kubelet doesn't do any feature gate checks, it just accepts from KAS and processes it, regardless of whether the feature is enabled in the kubelet or not

Reviewer 9: I'm guessing that's from a time when feature gates were not a thing. In which case I'd put an explicit information how this is currently implemented and update also `kep.yaml` since it's misleadingly pointing to both kube-apiserver and kubelet as having that feature gate.

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ProcMountType
  - Components depending on the feature gate: kubelet, kube-apiserver (whole cluster must have `ProcMountType` feature gate enabled)

###### Does enabling the feature change any default behavior?

No, only allows a user to access the Unmasked ProcMountType

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. This can be done by removing the feature gate from all kube-apiservers, and then restarting the nodes (to restart the pods that originally got
access to Unmasked ProcMountType).

###### What happens if we reenable the feature if it was previously rolled back?

Nothing. All pods that requested Unmasked ProcMountType had it stripped from the request because the feature was off. All other workloads remain unchanged.
Recreating pods that requested Unmasked will result in it being applied as expected, once the feature is on again.

###### Are there any tests for feature enablement/disablement?

> No

Reviewer 9: I'd say those are required for the beta promotion. We need to make sure that turning it off works as expected.

Author: I have left a comment describing my position, PTAL

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

Inconsistent feature gate enablement is the only way. All supported versions of CRIs (CRI-O and containerd) have support for this feature.

###### What specific metrics should inform a rollback?

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

> They will be, specifically cases where the Kubelet is restarted.

Reviewer 9: This answer should describe the manual testing scenario performed to ensure the path was tested.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

`kubectl get pods --all-namespaces -o jsonpath="{range .items[*]}{.metadata.name}{' '}{.spec.containers[*].securityContext.procMount}{'\n'}{end}"  | grep -i unmasked`
Will print all pods that has an Unmasked ProcMountType, along with the pod name.

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [x] Other (treat as last resort)
  - Details: Container created with the Unmasked ProcMountType have paths [here](https://github.com/kubernetes/kubernetes/blob/964529b/pkg/securitycontext/util.go#L193) as writable, not read only.
    - "/proc/asound",
	- "/proc/acpi",
	- "/proc/kcore",
	- "/proc/keys",
	- "/proc/latency_stats",
	- "/proc/timer_list",
	- "/proc/timer_stats",
	- "/proc/sched_debug",
	- "/proc/scsi",
	- "/sys/firmware",

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

> N/A

Reviewer 9: Don't we have a reasonable SLO for that kubelet_pod_start_sli_duration_seconds metric? Something like a few seconds? At least you can describe that the metric should maintain values from before the upgrade, for example.

Reviewer 9: see https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#kubernetes-slisslos

Author: updated!

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [x] Metrics
  - Metric name: `kubelet_pod_start_sli_duration_seconds`
  - [Optional] Aggregation method:
  - Components exposing the metric: kubelet
- [ ] Other (treat as last resort)
  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

> 

Reviewer 9: Missing answer here.

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

- A CRI implementation that supports this feature
    - All supported versions currently do.

### Scalability

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

ProcMountType in the pod spec

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

> Negligible

Reviewer 9: Please be more specific wrt to the size, see the template:
```
Describe them, providing:
  - API type(s):
  - Estimated increase in size: (e.g., new annotation of size 32B)
  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
 ```

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ProcMountType
  - Components depending on the feature gate: kube-apiserver (kube-apiserver filters `procMount` field if it's not enabled).

###### Does enabling the feature change any default behavior?

No, only gives a user access to the Unmasked ProcMountType

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. This can be done by removing the feature gate from all kube-apiservers. To fully roll back, the nodes will need to be drained or rebooted,
as the Kubelet will not remove the `procMount` of an already running container.

###### What happens if we reenable the feature if it was previously rolled back?

Nothing special. The pod's `procMount` field depends on where in the enablement process the kube-apiserver was when it was created.
The container has to be restarted to be up to date with the kube-apiserver.

###### Are there any tests for feature enablement/disablement?

Yes. I have manually tested feature enablement and disablement on kube-apiserver, and verified that pods are not recreated without

> a drain. There will be an e2e test to verify this as well.

Reviewer: if this ends up impractical in practice a unit test showing the field stripping will be sufficient along with a manual test of disablement.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ProcMountType
  - Components depending on the feature gate: kube-apiserver (kube-apiserver filters `procMount` field if it's not enabled).

###### Does enabling the feature change any default behavior?

No, only gives a user access to the Unmasked ProcMountType

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. This can be done by removing the feature gate from all kube-apiservers. To fully roll back, the nodes will need to be drained or rebooted,
as the Kubelet will not remove the `procMount` of an already running container.

###### What happens if we reenable the feature if it was previously rolled back?

Nothing special. The pod's `procMount` field depends on where in the enablement process the kube-apiserver was when it was created.
The container has to be restarted to be up to date with the kube-apiserver.

###### Are there any tests for feature enablement/disablement?

> There are not. I'm unsure it will be useful, as the toggling of enablement and the functional enablement of old containers are not related.

Reviewer: This must be manually tested to ensure that what is documented in https://github.com/kubernetes/enhancements/pull/4266/files#diff-1d1320b7f2cd6bb7e0cb6ebf236babce8b5bb99eaac25bd7fea57d1c17f21b52R348 functions as expected if it must be used in the field.

Author: updated to describe the results of a manual test, and describe plans to add a test
